nav2_collision_monitor/include/nav2_collision_monitor/source_base.hpp

```diff
@@ -0,0 +1,133 @@
+    std::shared_ptr<tf2_ros::Buffer> tf_buffer,
+    const std::string & source_name,
+    const std::string & base_frame_id,
+    const tf2::Duration & transform_tolerance,
+    const tf2::Duration & data_timeout);
+  /**
+   * @brief SourceBase destructor
+   */
+  virtual ~SourceBase();
+
+  /**
+   * @brief Data source configuration routine.
+   * Empty virtual method intended to be used in child implementations.
+   */
+  virtual void configure() = 0;
+
+  /**
+   * @brief Sets fixed frame ID. Used in data frame interpolation
+   * @param fixed_frame_id Fixed frame ID
+   */
+  void setFixedFrameId(const std::string & fixed_frame_id);
+
+  /**
+   * @brief Adds latest data from source to the data array.
+   * Empty virtual method intended to be used in child implementations.
+   * @param curr_time Current node time for data interpolation
+   * @param data Array where the data from source to be added.
+   * Added data is converted to base_frame_id coordinate system at curr_time.
+   */
+  virtual void getData(
+    const rclcpp::Time & curr_time, std::vector<Point> & data) = 0;
+
+protected:
+  /**
+   * @brief Supporting routine obtaining all ROS-parameters
+   * @param source_topic Output name of source subscription topic
+   */
+  void getParameters(std::string & source_topic);
+
+  /**
+   * @brief Checks whether the source might be considered as valid
+   * @param curr_time Current node time for source verification
+   * @return True if data source is valid, otherwise false
+   */
+  bool sourceValid(const rclcpp::Time & curr_time);
+
+  /**
+   * @brief Obtains a transform from source_frame_id at source_time ->
+   * to base_frame_id at current node time
+   * @param curr_time Current node timestamp to interpolate data to
+   * @param source_time Source timestamp to convert data from
+   * @param tf2_transform Output source->base transform
+   * @return True if got correct transform, otherwise false
+   */
+  bool getSourceBaseTransform(
+    const rclcpp::Time & curr_time,
+    const rclcpp::Time & source_time,
+    tf2::Transform & tf2_transform);
+
+  // ----- Variables -----
+
+  /// @brief Collision Monitor node
+  nav2_util::LifecycleNode::WeakPtr node_;
+  /// @brief Collision monitor node logger stored for further usage
+  rclcpp::Logger logger_{rclcpp::get_logger("collision_monitor")};
+
+  /// @brief TF buffer
+  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
+
+  // Basic parameters
+  /// @brief Name of data source
+  std::string source_name_;
+  /// @brief Source frame ID
+  std::string source_frame_id_;
+  /// @brief Base frame ID
+  std::string base_frame_id_;
+  /// @brief Fixed frame ID. Used to get source->base time inerpolated transform.
+  std::string fixed_frame_id_;
+
+  /// @brief Transform tolerance
+  tf2::Duration transform_tolerance_;
+  /// @brief Maximum allowed time shift between data and collision monitor node
+  tf2::Duration data_timeout_;
+
+  /// @brief Latest data timestamp
+  rclcpp::Time data_stamp_;
```

Comment:
`data_` in PC2 and LS have a `header.stamp` field -- why not use that?

Reply:
This parameter is [used](https://github.com/ros-planning/navigation2/pull/2982/files#diff-7e24f9dbbee0e03a22353c8e2625792fc824ced725395a85665629f232ce4706R71) by parent `SourceBase` class, which has no knowledge about type of `data_` in child ones. Therefore, I forced to do this variable to store value in childs for usage in parent' method.

Reply:
I think you mentioned somewhere else (that I can't find again) that you were concerned since the base class wouldn't know about this. You could just make `SourceBase<>` a template on the message type so that it stores a `T msg_` as its most recent value containing the time / data. Then you also wouldn't need a `data_` in the class implementations of the sources and could even make the subscriber a member of the `source base`. That would further simplify the scan/pointcloud to just `getData` / `getParameters` / `dataCallback` implementations.

Reply:
Also, you could remove `source_frame_id_`, since the `T msg_` would contain it just like the timestamp.

Reply:
Currently, it is passed as follows:
```
 /**
   * @brief Checks whether the source data might be considered as valid
   * @param source_time Timestamp of latest obtained data
   * @param curr_time Current node time for source verification
   * @param data_timeout Maximum time interval in which data is considered valid
   * @return True if data source is valid, otherwise false
   */
  bool sourceValid(
    const rclcpp::Time & source_time,
    const rclcpp::Time & curr_time,
    const rclcpp::Duration & data_timeout) const;
```

From my point of view - this method is already quite compact. However, your idea with templating this function is also works fine. Do you consider we need to make a templates it to avoid one argument in this function?

Reply:
It was more about simplifying a bunch of stuff at once (e.g. the `data_` can be stored in `source` as well as the subscriber) but no, I don't think it matters that much as long as we've removed the duplicate and spread out information from a compact single source.